File: vscode/src/core/traitPresets.ts
import { TRAIT_CATALOG } from './traitCatalog';
import { TraitIds } from './traitIds';
import { TraitLayer } from './traitLayer';

export interface TraitPreset {
    readonly id: string;
    readonly name: string;
    readonly description: string;
    readonly weights: Readonly<Record<string, number>>;
}

/**
 * Divides a budget evenly across every trait on one layer.
 *
 * Layers other than Effect share a single 0-100 roll, so their weights have to sum to the
 * budget rather than each being an independent chance. Spreading rather than setting each
 * to the budget is what stops the first trait in the list swallowing every brace.
 */
function layerSpread(layer: TraitLayer, budget: number): Record<string, number> {
    const members = TRAIT_CATALOG.filter((t) => t.layer === layer);
    const weights: Record<string, number> = {};
    if (members.length === 0) {
        return weights;
    }

    const each = Math.max(1, Math.trunc(budget / members.length));
    for (const member of members) {
        weights[member.id] = each;
    }

    return weights;
}

function merge(...parts: Record<string, number>[]): Record<string, number> {
    return Object.assign({}, ...parts);
}

function menagerie(): Record<string, number> {
    const weights = merge(layerSpread(TraitLayer.Creature, 70), layerSpread(TraitLayer.Costume, 55));
    weights[TraitIds.Question] = 5;
    weights[TraitIds.ColourCycle] = 10;
    return weights;
}

function unusable(): Record<string, number> {
    const weights = merge(
        layerSpread(TraitLayer.Creature, 85),
        layerSpread(TraitLayer.Costume, 85),
        layerSpread(TraitLayer.Motion, 85),
        layerSpread(TraitLayer.Body, 45),
    );

    weights[TraitIds.Fire] = 8;
    weights[TraitIds.Tilted] = 25;
    weights[TraitIds.Shadow] = 20;
    weights[TraitIds.GradientFill] = 20;
    weights[TraitIds.BoldItalic] = 20;
    weights[TraitIds.Underline] = 10;
    weights[TraitIds.Distressed] = 12;
    return weights;
}

export const TRAIT_PRESETS: readonly TraitPreset[] = [
    {
        id: 'off',
        name: 'Off',
        description: 'Colours only. Every brace keeps its own glyph.',
        weights: {},
    },
    {
        id: 'default',
        name: 'Default',
        description:
            'What the extension shipped with: a few questions, some colour cycling, the occasional cat.',
        weights: {
            [TraitIds.Question]: 6,
            [TraitIds.ColourCycle]: 8,
            [TraitIds.Catgirl]: 4,
            [TraitIds.ThighHighs]: 4,
        },
    },
    {
        id: 'menagerie',
        name: 'Menagerie',
        description: 'Creatures and costumes across the board, but nothing that moves much.',
        weights: menagerie(),
    },
    {
        id: 'restless',
        name: 'Restless',
        description: 'Everything on the motion layer, spread evenly. Nothing on screen holds still.',
        weights: layerSpread(TraitLayer.Motion, 90),
    },
    {
        id: 'unusable',
        name: 'Unusable',
        description: 'Every layer saturated. This is the setting the extension was designed for.',
        weights: unusable(),
    },
];

export function findPreset(id: string): TraitPreset | undefined {
    return TRAIT_PRESETS.find((p) => p.id === id);
}

/**
 * A preset expressed as the full override map the settings store keeps.
 *
 * Every trait is listed explicitly, including the zeroes: a preset that only wrote its
 * non-zero entries would leave whatever the previous preset had turned on still running.
 */
export function presetWeights(preset: TraitPreset): Record<string, number> {
    const weights: Record<string, number> = {};
    for (const trait of TRAIT_CATALOG) {
        weights[trait.id] = preset.weights[trait.id] ?? 0;
    }

    return weights;
}

File: jetbrains/src/main/kotlin/identitybraces/core/TraitPresets.kt
package identitybraces.core

/** A named set of trait weights. Anything not listed in [weights] is zero. */
class TraitPreset(val id: String, val name: String, val description: String, val weights: Map<String, Int>)

/**
 * Bundles that set every weight at once.
 *
 * Eighty-six individual percentages is not a settings page, it is a spreadsheet. A preset is
 * how someone gets a coherent look without reading the whole catalogue, and picking one is
 * still just a write to the same weight table — so a preset can be applied and then
 * adjusted, rather than being a mode you are locked into.
 */
object TraitPresets {
    val all: List<TraitPreset> = listOf(
        TraitPreset("off", "Off", "Colours only. Every brace keeps its own glyph.", emptyMap()),
        TraitPreset(
            "default", "Default",
            "What the extension shipped with: a few questions, some colour cycling, the occasional cat.",
            mapOf(
                TraitIds.QUESTION to 6,
                TraitIds.COLOUR_CYCLE to 8,
                TraitIds.CATGIRL to 4,
                TraitIds.THIGH_HIGHS to 4,
            ),
        ),
        TraitPreset(
            "menagerie", "Menagerie",
            "Creatures and costumes across the board, but nothing that moves much.",
            buildMenagerie(),
        ),
        TraitPreset(
            "restless", "Restless",
            "Everything on the motion layer, spread evenly. Nothing on screen holds still.",
            layerSpread(TraitLayer.Motion, 90),
        ),
        TraitPreset(
            "unusable", "Unusable",
            "Every layer saturated. This is the setting the extension was designed for.",
            buildUnusable(),
        ),
    )

    fun find(id: String): TraitPreset? = all.firstOrNull { it.id == id }

    /**
     * A preset expressed as the full override map the settings store keeps.
     *
     * Every trait is listed explicitly, including the zeroes: a preset that only wrote its
     * non-zero entries would leave whatever the previous preset had turned on still running.
     */
    fun weightsOf(preset: TraitPreset): Map<String, Int> {
        val weights = LinkedHashMap<String, Int>()
        for (trait in TraitCatalog.all) {
            weights[trait.id] = preset.weights[trait.id] ?: 0
        }

        return weights
    }

    private fun buildMenagerie(): Map<String, Int> {
        val weights = LinkedHashMap<String, Int>()
        weights.putAll(layerSpread(TraitLayer.Creature, 70))
        weights.putAll(layerSpread(TraitLayer.Costume, 55))
        weights[TraitIds.QUESTION] = 5
        weights[TraitIds.COLOUR_CYCLE] = 10
        return weights
    }

    private fun buildUnusable(): Map<String, Int> {
        val weights = LinkedHashMap<String, Int>()
        weights.putAll(layerSpread(TraitLayer.Creature, 85))
        weights.putAll(layerSpread(TraitLayer.Costume, 85))
        weights.putAll(layerSpread(TraitLayer.Motion, 85))
        weights.putAll(layerSpread(TraitLayer.Body, 45))

        // Effects roll independently rather than sharing a band, so they are set per trait
        // rather than spread across a budget.
        weights[TraitIds.FIRE] = 8
        weights[TraitIds.TILTED] = 25
        weights[TraitIds.SHADOW] = 20
        weights[TraitIds.GRADIENT_FILL] = 20
        weights[TraitIds.BOLD_ITALIC] = 20
        weights[TraitIds.UNDERLINE] = 10
        weights[TraitIds.DISTRESSED] = 12
        return weights
    }

    /**
     * Divides a budget evenly across every trait on one layer.
     *
     * Layers other than Effect share a single 0–100 roll, so their weights have to sum to
     * the budget rather than each being an independent chance. Spreading rather than setting
     * each to the budget is what stops the first trait in the list swallowing every brace.
     */
    private fun layerSpread(layer: TraitLayer, budget: Int): Map<String, Int> {
        val members = TraitCatalog.all.filter { it.layer == layer }
        if (members.isEmpty()) {
            return emptyMap()
        }

        val each = maxOf(1, budget / members.size)
        val weights = LinkedHashMap<String, Int>()
        for (member in members) {
            weights[member.id] = each
        }

        return weights
    }
}
